riscv-sbi.md

```diff
@@ -0,0 +1,113 @@
+# RISC-V SBI specification
+
+## Table of Contents
+1. [Introduction](#Introduction)
+2. [SBI List](#sbi-list)
+3. [Description](#Description)
+	* [Timer](#Timer)
+	* [Inter Processor Interrupt (IPI)](#IPI)
+	* [Memory Model](#memory-model)
+	* [Console](#Console)
+	* [Shutdown](#Shutdown)
+4. [Conclusion](#Conclusion)
+
+
+## Copyright and license information
+
+This RISC-V Supervisor Binary Interface(SBI) specification document is
+
+ &copy; 2018 Atish Patra <atish.patra@wdc.com>
+ &copy; 2018 Andrew Waterman <aswaterman@gmail.com>
+ &copy; 2018 Palmer Dabbelt <palmer@dabbelt.com>
+
+It is licensed under the Creative Commons Attribution 4.0 International
+License (CC-BY 4.0).  The full license text is available at
+https://creativecommons.org/licenses/by/4.0/.
+
+## Introduction
+
+Supervisor Binary Interface(SBI) describes the instructions that can be executed together with a set of SBI calls out to the supervisor execution environment (SEE)
+on a given platform. Several instructions that might normally be handled by the
+supervisor operating system (OS) directly are handled via SBI calls to the SEE in
+RISC-V. This provides a cleaner interface for the supervisor OS. Currently,
+The next section will describe different types of SBI calls defined for RISC-V.
+
+## SBI List<a name=sbi-list>
+There are mainly 5 types of SBI calls defined in the specification as per below
+table.
+
+| Type          | Function          |
+| --------------|-------------------|
+| Timer         | sbi_set_timer     |
+| IPI           | sbi_clear_ipi<br>sbi_send_ipi  |
+| Memory Model| sbi_remote_fence_i<br>sbi_remote_sfence_vma<br>sbi_remote_sfence_vma_asid
+|	Console				| sbi_console_putchar <br> sbi_console_getchar |
+| Shutdown         |	sbi_shutdown |
+
+## Description
+Each SBI call adheres to a specific calling convention that defines how data is provided as input to or read as output. Individual SBI call signature and its
+purpose is described next.
+
+### Timer
+```C
+void sbi_set_timer(uint64_t stime_value)
+```
+Programs the clock for next event after *stime_value* time. This function also
+clears the pending timer interrupt bit. If the supervisor needs to clear the
+timer interrupt without scheduling the next timer event, they may choose to do
+so in timer interrupt handler or any other place in supervisor.
+
+### IPI
+```C
+void sbi_send_ipi(const unsigned long *hart_mask)
+```
+Send an inter-processor interrupt to all the harts defined in hart_mask.
```

Comment:
Suggest adding the following text:

"Interprocessor interrupts manifest at the receiving harts as Supervisor Software Interrupts.

hart_mask is a physical address that points to a bit-vector of harts.  The bit vector is represented as a sequence of unsigned longs whose length equals the number of harts in the system divided by the number of bits in an unsigned long, rounded up to the next integer.

If hart_mask is NULL, an IPI is sent to every hart in the system."

Reply:
Sure.